### PR TITLE
Crimap 219 v2 search by name and reference

### PR DESCRIPTION
**Pull request by timpeat**
## Description of change
Adds searchable column for text based query.
Currently indexes only first name, last name, and reference.

## Link to relevant ticket
[CRIMRE-83](https://dsdmoj.atlassian.net/browse/CRIMRE-83)

### CODE REVIEW

File: app/api/datastore/entities/search_result.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    class SearchResult < Grape::Entity
+      expose :id, as: :resource_id
+      expose :submitted_at
+      expose :applicant_name
+      expose :reference
+
+      private
+
+      def applicant_name
+        "#{applicant&.dig('first_name')} #{applicant&.dig('last_name')}"
```
Comment (zheileman) [inline app/api/datastore/entities/search_result.rb:12]: This really smells like DynamoDB all over again 😄 

In DynamoDB the idea was to have hash keys, `first_name#last_name#dob` etc.

As you advance, keep an eye to see if in reality everything that we are doing (apart from pagination) with Postgres might not be better suited with Dynamo 👍 
Pagination in an extreme situation I think could be handled as well...

**Review by willmcb — CHANGES_REQUESTED** [review]
Nice one, few small points :)

File: db/schema.rb [hunk 2]
```diff
@@ -21,8 +21,8 @@
     t.text "status", default: "submitted", null: false
     t.datetime "submitted_at", precision: nil, default: -> { "CURRENT_TIMESTAMP" }
     t.datetime "returned_at", precision: nil
-    t.index ["status", "returned_at"], name: "index_crime_applications_on_status_and_returned_at", order: { returned_at: :desc }
-    t.index ["status", "submitted_at"], name: "index_crime_applications_on_status_and_submitted_at", order: { submitted_at: :desc }
+    t.virtual "searchable_text", type: :tsvector, as: "((to_tsvector('english'::regconfig, (application #>> '{client_details,applicant,first_name}'::text[])) || to_tsvector('english'::regconfig, (application #>> '{client_details,applicant,last_name}'::text[]))) || to_tsvector('english'::regconfig, (application ->> 'reference'::text)))", stored: true
```
[the remark this answers is not in the document]
Comment (timpeat) [reply db/schema.rb:24]: Fixed.

**Review by willmcb — APPROVED** [review]
LGTM 👍

File: app/api/datastore/entities/search_result.rb [hunk 3]
```diff
@@ -2,6 +2,23 @@ module Datastore
...
     class SearchResult < Grape::Entity
       expose :id, as: :resource_id
+      expose :submitted_at
+      expose :applicant_name
+      expose :reference
+
+      private
+
+      def applicant_name
+        "#{applicant&.dig('first_name')} #{applicant&.dig('last_name')}"
```
Comment (arthurashman) [inline app/api/datastore/entities/search_result.rb:12]: I _think_ .dig has the null safety baked into it

**Review by arthurashman — APPROVED** [review]

File: app/api/datastore/entities/search_result.rb (hunk 3 above)
Comment (timpeat) [reply app/api/datastore/entities/search_result.rb:12]: But not if the hash itself is null.

Comment (arthurashman) [reply app/api/datastore/entities/search_result.rb:12]: Aha, ok 👌

**Pull request merged**